lib/internal/modules/esm/hooks.js

```diff
@@ -516,9 +460,187 @@ class Hooks {
+      this.worker.terminate();
+    } else {
+      // The worker is still in good health, and we just told it it's the
+      // only thing keeping us alive.
+      this.worker.ref();
+    }
+  };
+
+  async makeAsyncRequest(method, ...args) {
+    AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_MESSAGES, 1);
+    this.#waitForWorker();
+
+    MessageChannel ??= require('internal/worker/io').MessageChannel;
+    const asyncCommChannel = new MessageChannel();
+
+    // Pass work to the worker.
+    debug('post async message to worker', { method, args });
+    this.#syncCommPort.postMessage({ method, args, port: asyncCommChannel.port2 }, [asyncCommChannel.port2]);
+    // In case the worker was bored, it might lock itself waiting for a signal.
+    AtomicsNotify(this.#lock, NUMBER_OF_INCOMING_MESSAGES);
+
+    if (AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_ASYNC_RESPONSES, 1) === 0) {
+      process.on('beforeExit', this.#beforeExitHandler);
+    }
+
+    let response;
+    debug('start waiting for async response from worker', { method, args });
+    while (response == null) {
+      await AtomicsWaitAsync(this.#lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 0).value;
+      AtomicsAnd(this.#lock, IDLENESS_OF_THREADS, ~MAIN_THREAD_IS_IDLE); // We are no longer bored.
+
+      response = receiveMessageOnPort(asyncCommChannel.port1);
+    }
+    debug('got async response from worker', { method, args }, this.#lock);
+
+    if (AtomicsSub(this.#lock, NUMBER_OF_INCOMING_ASYNC_RESPONSES, 1) <= 1) {
+      // We got all the responses from the worker, its job is done (until next time).
+      this.worker.unref();
+      process.off('beforeExit', this.#beforeExitHandler);
+    }
+
+    const body = this.#unwrapMessage(response);
+    asyncCommChannel.port1.close();
+    return body;
+  }
+
+  makeSyncRequest(method, ...args) {
+    AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_MESSAGES, 1);
+    this.#waitForWorker();
+
+    // Pass work to the worker.
+    debug('post sync message to worker', { method, args });
+    this.#syncCommPort.postMessage({ method, args });
+
+    // In case the worker was bored, it might lock itself waiting for a signal.
+    AtomicsNotify(this.#lock, NUMBER_OF_INCOMING_MESSAGES);
+
+    let response;
+    debug('start waiting for sync response from worker', { method, args });
+    while (response == null) {
+      AtomicsWait(this.#lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 0); // Sleep until worker responds.
+
+      response = receiveMessageOnPort(this.#syncCommPort);
+
+      if (response == null && AtomicsLoad(this.#lock, IDLENESS_OF_THREADS) & WORKER_THREAD_IS_IDLE) {
+        // TODO(aduh95): handle never-settling promises in sync responses.
+      }
+    }
+    debug('got sync response from worker', { method, args });
+    return this.#unwrapMessage(response);
+  }
+
+  #unwrapMessage(response) {
+    if (response.message.status === 'error') {
+      if (response.message.body == null) throw response.message.body;
```

Comment:
throw nullish?

Reply:
Exactly

Reply:
I can see it's throwing nullish. won't that mean the error message is empty?

Reply:
It means the loader thread has thrown some nullish value, so there's nothing to reserialize and we can throw it immediately. The error message is not empty, there's no error message at all.